src/common/cups_print.c

```diff
@@ -323,61 +397,130 @@ void dt_print_file(const int32_t imgid, const char *filename, const dt_print_inf
     return;
   }
 
-  cups_dest_t *dests;
-  int num_dests = cupsGetDests(&dests);
-  cups_dest_t *dest = cupsGetDest(pinfo->printer.name, NULL, num_dests, dests);
-
   cups_option_t *options = NULL;
   int num_options = 0;
 
-  for (int j = 0; j < dest->num_options; j ++)
-    if (cupsGetOption(dest->options[j].name, num_options,
-                      options) == NULL)
-      num_options = cupsAddOption(dest->options[j].name,
-                                  dest->options[j].value,
-                                  num_options, &options);
+  // for turboprint drived printer, use the turboprint dialog
+  if (pinfo->printer.is_turboprint)
+  {
+    const char *tp_intent_name[] = { "perception_0", "colorimetric-relative_1", "saturation_1", "colorimetric-absolute_1" };
+    char tpcmd[1024];
+    char tmpfile[PATH_MAX] = { 0 };
 
-  cupsFreeDests(num_dests, dests);
+    dt_loc_get_tmp_dir(tmpfile, sizeof(tmpfile));
+    g_strlcat(tmpfile, "/dt_cups_opts_XXXXXX", sizeof(tmpfile));
 
-  // if we have a profile, disable cm on CUPS, this is important as dt does the cm
+    gint fd = g_mkstemp(tmpfile);
+    if(fd == -1)
+    {
+      dt_control_log(_("failed to create temporary file for printing options"));
+      fprintf(stderr, "failed to create temporary pdf for printing options\n");
+      return;
+    }
+    close(fd);
 
-  num_options = cupsAddOption("cm-calibration", *pinfo->printer.profile ? "true" : "false", num_options, &options);
+    // ensure that intent is in the range, may happen if at some point we add new intent in the list
+    const int intent = (pinfo->printer.intent < 4) ? pinfo->printer.intent : 0;
 
-  // media to print on
+    // start the turboprint dialog
+    snprintf
+      (tpcmd, sizeof(tpcmd),
+       "turboprint --printer=%s --options --output=%s -o copies=1 -o PageSize=%s -o InputSlot=AutoSelect -o zedoIntent=%s -o MediaType=%s",
+       pinfo->printer.name, tmpfile, pinfo->paper.common_name, tp_intent_name[intent], pinfo->medium.name);
 
-  num_options = cupsAddOption("media", pinfo->paper.name, num_options, &options);
+    dt_print(DT_DEBUG_PRINT, "[print]   cmd='%s'\n", tpcmd);
 
-  // never print two-side
+    const int res = system(tpcmd);
```

Comment:
Welcome to the wonderful world of SQL injections.

Reply:
Not sure about SQL injections since this is a system command. What you would do?

Reply:
https://codereview.stackexchange.com/questions/6435/c-system-function-vulnerability
https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152177

Need more links?

It appears we already have a few `system(3)` calls, and this is disgusting :(

Reply:
No I don't need more link. I know that a system call is not safe but I'm questioning the SQL injections. And yes we already have some system() calls. darktable is not a highly secure embedded control code for whatever life critical application :)

Reply:
e.g. `pinfo->printer.name` roughly of `; rm -rf ~/`, when printed to, will delete your homedir.
You think this is ok?

Reply:
No it is not, of course and I'm aware of that. Many applications are using system() or exec*() to spawn external process. We do in different places and have no alternatives, do we?

Reply:
Relevant link: https://xkcd.com/327/

> Many applications are using system() or exec*() to spawn external process. 

.. many applications leak memory, many applications contain UB, etc.
This isn't false, it just doesn't mean that because of that, it is good to follow the trend.

> and have no alternatives, do we?
Or do we not?

```
SYSTEM(3)                                                                                                               Linux Programmer's Manual                                                                                                              SYSTEM(3)

NAME
       system - execute a shell command

SYNOPSIS
       #include <stdlib.h>

       int system(const char *command);

DESCRIPTION
       The system() library function uses fork(2) to create a child process that executes the shell command specified in command using execl(3) as follows:

           execl("/bin/sh", "sh", "-c", command, (char *) 0);
```

Basically we just need to remove the `"/bin/sh", "sh", "-c", ` from the equation, that's it.
I.e. we know what command we want to call, so let's call it, and not use shell to call it.

Reply:
AFAICT there is only one `system` call in dt that counts. The one from Lua is just wrapping what the user feeds it. And the rest it test code or external libs.

That being said, just use g_spawn_* as it's done in `src/views/lighttable.c`, maybe sync, maybe async.